log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -0,0 +1,579 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Handler;
+import java.util.logging.LogRecord;
+import java.util.zip.GZIPOutputStream;
+
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.lang.Math.toIntExact;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static java.util.logging.ErrorManager.CLOSE_FAILURE;
+import static java.util.logging.ErrorManager.FLUSH_FAILURE;
+import static java.util.logging.ErrorManager.FORMAT_FAILURE;
+import static java.util.logging.ErrorManager.GENERIC_FAILURE;
+import static java.util.logging.ErrorManager.OPEN_FAILURE;
+import static java.util.logging.ErrorManager.WRITE_FAILURE;
+
+public final class RollingFileHandler
+        extends Handler
+{
+    public enum CompressionType
+    {
+        NONE(Optional.empty()),
+        GZIP(Optional.of(".gz"));
+
+        private final Optional<String> extension;
+
+        CompressionType(Optional<String> extension)
+        {
+            this.extension = requireNonNull(extension, "extension is null");
+        }
+
+        public Optional<String> getExtension()
+        {
+            return extension;
+        }
+    }
+
+    private static final int MAX_BATCH_COUNT = 1024;
+    private static final int MAX_BATCH_BYTES = toIntExact(new DataSize(1, MEGABYTE).toBytes());
+
+    private static final String TEMP_PREFIX = ".tmp.";
+    private static final String DELETED_PREFIX = ".deleted.";
+
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("-yyyyMMdd.HHmmss");
+
+    private final Path link;
+    private final long maxFileSize;
+    private final CompressionType compressionType;
+
+    @GuardedBy("this")
+    private Path currentOutputFile;
+    @GuardedBy("this")
+    private LogFileName currentOutputFileName;
+    @GuardedBy("this")
+    private OutputStream currentOutputStream;
+    @GuardedBy("this")
+    private long currentFileSize;
+    private final LogHistoryManager logHistoryManager;
```

Comment:
I'd name this just `historyManager`. It's understood that it's for logs from the context.